From: Alberto Garcia <berto@igalia.com>
Subject: Fix the MiniBrowser build ("error: expected expression")
Forwarded: no
Index: webkitgtk/Tools/MiniBrowser/gtk/main.c
===================================================================
--- webkitgtk.orig/Tools/MiniBrowser/gtk/main.c
+++ webkitgtk/Tools/MiniBrowser/gtk/main.c
@@ -288,10 +288,8 @@ static gboolean parseFeaturesOptionCallb
         case '!':
         case '-':
             enabled = FALSE;
-            [[fallthrough]];
         case '+':
             item++;
-            [[fallthrough]];
         default:
             break;
         }
