From: Alberto Garcia <berto@igalia.com>
Subject: Fix FTBFS in Hurd
Forwarded: no
Index: webkitgtk/Source/WTF/wtf/InlineASM.h
===================================================================
--- webkitgtk.orig/Source/WTF/wtf/InlineASM.h
+++ webkitgtk/Source/WTF/wtf/InlineASM.h
@@ -82,6 +82,7 @@
 #elif   OS(LINUX)               \
      || OS(FREEBSD)             \
      || OS(OPENBSD)             \
+     || OS(HURD)                \
      || OS(NETBSD)
     // GNU as-compatible syntax.
 #define LOCAL_LABEL_STRING(name) ".L" #name
