From: Andreas Kling  <akling@apple.com>
Subject: Fix inserted text when typing in the Twitter message box.
Origin: http://trac.webkit.org/changeset/176824
Index: webkitgtk/Source/JavaScriptCore/runtime/JSString.h
===================================================================
--- webkitgtk.orig/Source/JavaScriptCore/runtime/JSString.h
+++ webkitgtk/Source/JavaScriptCore/runtime/JSString.h
@@ -482,12 +482,7 @@ ALWAYS_INLINE AtomicStringImpl* JSString
         return static_cast<const JSRopeString*>(this)->resolveRopeToExistingAtomicString(exec);
     if (m_value.impl()->isAtomic())
         return static_cast<AtomicStringImpl*>(m_value.impl());
-    if (AtomicStringImpl* existingAtomicString = AtomicString::find(m_value.impl())) {
-        m_value = *existingAtomicString;
-        setIs8Bit(m_value.impl()->is8Bit());
-        return existingAtomicString;
-    }
-    return nullptr;
+    return AtomicString::find(m_value.impl());
 }
 
 inline const String& JSString::value(ExecState* exec) const
